evidence/reactor_test.go

```diff
@@ -31,129 +33,172 @@ var (
+		blockStore := &mocks.BlockStore{}
+		blockStore.On("LoadBlockMeta", mock.AnythingOfType("int64")).Return(
+			&types.BlockMeta{Header: types.Header{Time: evidenceTime}},
+		)
+		rts.pools[nodeID], err = evidence.NewPool(logger, evidenceDB, stateStores[idx], blockStore)
+
+		require.NoError(t, err)
+
+		rts.peerChans[nodeID] = make(chan p2p.PeerUpdate)
+		rts.peerUpdates[nodeID] = p2p.NewPeerUpdates(rts.peerChans[nodeID])
+		rts.network.Nodes[nodeID].PeerManager.Register(rts.peerUpdates[nodeID])
+		rts.nodes = append(rts.nodes, rts.network.Nodes[nodeID])
 
-	require.NoError(t, rts.reactor.Start())
-	require.True(t, rts.reactor.IsRunning())
+		rts.reactors[nodeID] = evidence.NewReactor(logger,
+			rts.evidenceChannels[nodeID],
+			rts.peerUpdates[nodeID],
+			rts.pools[nodeID])
+
+		require.NoError(t, rts.reactors[nodeID].Start())
+		require.True(t, rts.reactors[nodeID].IsRunning())
+
+		idx++
+	}
 
 	t.Cleanup(func() {
-		require.NoError(t, rts.reactor.Stop())
-		require.False(t, rts.reactor.IsRunning())
+		for _, r := range rts.reactors {
+			if r.IsRunning() {
+				require.NoError(t, r.Stop())
+				require.False(t, r.IsRunning())
+			}
+		}
+
+		leaktest.Check(t)
 	})
 
 	return rts
 }
 
-func createTestSuites(t *testing.T, stateStores []sm.Store, chBuf uint) []*reactorTestSuite {
-	t.Helper()
+func (rts *reactorTestSuite) start(t *testing.T) {
+	rts.network.Start(t)
+	require.Len(t, rts.network.RandomNode().PeerManager.Peers(), rts.numStateStores-1,
+		"network does not have expected number of nodes")
+}
 
-	numSStores := len(stateStores)
-	testSuites := make([]*reactorTestSuite, numSStores)
-	evidenceTime := time.Date(2019, 1, 1, 0, 0, 0, 0, time.UTC)
+func (rts *reactorTestSuite) waitForEvidence(t *testing.T, evList types.EvidenceList, ids ...p2p.NodeID) {
+	t.Helper()
 
-	for i := 0; i < numSStores; i++ {
-		logger := log.TestingLogger().With("validator", i)
-		evidenceDB := dbm.NewMemDB()
-		blockStore := &mocks.BlockStore{}
-		blockStore.On("LoadBlockMeta", mock.AnythingOfType("int64")).Return(
-			&types.BlockMeta{Header: types.Header{Time: evidenceTime}},
+	fn := func(pool *evidence.Pool) {
+		var (
+			localEvList []types.Evidence
+			size        int64
 		)
 
-		pool, err := evidence.NewPool(logger, evidenceDB, stateStores[i], blockStore)
-		require.NoError(t, err)
+		// wait till we have at least the amount of evidence
+		// that we expect. if there's more local evidence then
+		// it doesn't make sense to wait longer and a
+		// different assertion should catch the resulting error
+		loops := 0
+		for len(localEvList) < len(evList) {
+			// each evidence should not be more than 500 bytes
+			localEvList, size = pool.PendingEvidence(int64(len(evList) * 500))
+			if loops == 10 {
+				t.Log("current wait status:", "|",
+					"local", len(localEvList), "|",
+					"waitlist", len(evList), "|",
+					"size", size)
+			}
+			if loops == 10000 && size == 0 {
+				assert.Fail(t, "pool seems disconnected")
+				return
+			}
```

Comment:
This won't catch the case where we are waiting for say 10 pieces of evidence, but only 5 pieces of evidence are accepted by the evidence pool. It only covers when nothing makes its way to the pool.

Reply:
Ah, right. So I was ending up in situations where the `waitForEvidence`, would cause deadlocks, so this was an easier way of of just avoiding deadlocks in situations where we weren't making progress, and just makes the test easier to work with. 

I recognize that the diff is _particularly_ unclear in this situation, so [this is the version of this code in mainline](https://github.com/tendermint/tendermint/blob/master/evidence/reactor_test.go#L117-L157), and I think that these changes are an improvement, though clearly we could have better "waiting" semantics at some point.

Reply:
So when and where does this deadlock? I would think that having this logic less pragmatic, no? Is it that we're getting too much or not enough/expected evidence?

Reply:
I'm actually super not opposed to taking this out now that things are more stable (it was definitely useful during the refactor to make it easier to iterate. My thought is "if there's nothing in the pool, and we've been trying for a while, then its likely we're not making progress and it'd be better to abort rather than keep spinning," but if it's confusing, or my assumption is wrong, we should def just remove it.

Reply:
Typically, we just rely on the testing timeout rather than adding explicit timeout guards. That being said, there are cases where explicit guards should/can be used. But I don't think we need them here?